cli/azd/pkg/project/project_utils.go

```diff
@@ -48,17 +50,43 @@ type buildForZipOptions struct {
 	excludeConditions []excludeDirEntryCondition
 }
 
-// buildForZip is use by projects which build strategy is to only copy the source code into a folder which is later
-// zipped for packaging. For example Python and Node framework languages. buildForZipOptions provides the specific
-// details for each language which should not be ever copied.
-func buildForZip(src, dst string, options buildForZipOptions) error {
+// buildForZip is used by projects to prepare a directory for
+// zipping, excluding files based on the ignore file and other conditions.
+func buildForZip(src, dst string, options buildForZipOptions, serviceConfig *ServiceConfig) error {
+	// Lookup the appropriate ignore file name based on the service kind (Host)
+	ignoreFileName := GetIgnoreFileNameByKind(serviceConfig.Host)
+
+	// Read and honor the specified ignore file if it exists
+	ignoreMatcher, err := dotignore.ReadDotIgnoreFile(src, ignoreFileName)
```

Comment:
Are we also thinking about supporting `*.ignore` files in any recursive directory of the service?

```
- funcapp/
   - .funcignore
   - src/
     - otherThing/
       - .funcignore # define ignore files relative to this path
```

I think we would honor both ignore files.

If we're doing this, I think we should *consider* implementing the correct recursion logic in `project` before generalizing it as a `dotignore` package -- I would like to see the design matured out before we push it up to an app library package in `azd`.

Reply:
@ellismg had implemented `.dockerignore` recently -- wonder if he has thoughts here.

Reply:
> @ellismg had implemented `.dockerignore` recently

@ellismg - Are you thinking we migrate the pack_source.go implementation to use the new dotignore support in this PR?

Reply:
> .funcignore only honors files in the root and in discussion with them and the app service team we decided to not honor recursive .ignore files.

That is quite an interesting decision. From a dev perspective, one main reason why `*.ignore` files are used today is precisely because they can always be placed relative to a directory and provide a more local ignore scope. If that is not being done, a simpler package manifest at the root of the package directory would be a more ideal format IMO.

I took a closer look. It also turns out that `.funcignore` does not conform to the `.gitignore` spec entirely. @ellismg @jongio 

- The issue on [functools](https://github.com/Azure/azure-functions-core-tools/issues/2701)
- Based on the previous issue, I gather that the library they're using is [gitignore-parser](https://github.com/codemix/gitignore-parser/issues) with its set of issues.

If we're already engaged in conversations with both app service and function teams, can we consider:

1. Evaluating the impact of having two different `.funcignore` interpretations (is there user pain we would worry about?)
2. Based on those conversations, decide if we would like to rally all teams to adopt the conformant spec, and think about potentially support recursive directories. We could also just support recursive directories from `azd` side, it would be perhaps 20 lines of code.

Reply:
Clarifying that I meant to say..."service root", not the root of the repo.

This implementation honors .ignore files in the root of the service.  

I think we are on the same page.

Reply:
@weikanglim - Any objections to moving forward with the current implementation as it matches what .funcignore does today and what they would like for .webappignore.  

This means a service zipping only recognizes the .ignore file that is in the root of that service folder and not any parent .ignore files.

Reply:
@jongio Just checking -- where are we in resolving the difference between of behavior in interpretation for `.funcignore` by func-tools vs. azd, as highlighted [previously](https://github.com/Azure/azure-dev/pull/4258/files#r1767506732)? I haven't seen it mentioned.

Aside from that:
1. I'm a little unhappy about `dotignore` being a package and the implementation of the functions being tied to working "within a directory". I suspect in the future this may end up needing slight rewriting to support recursive directories. I'm more than happy to work with you to push some changes here to clean up the implementation.
2. I'm looking at the PR again, and it seems like there are significant changes to `rzip.go` that I'm unsure how or if they're related. They should be separated out if it's a bug fix.
3. I'm also a little nervous about the e2e tests that we're doing inspection of the zip files -- `package` likely requires changes in the future that would require us to rethink these test cases. The heavy addition of many test data files to set up "scenario tests" of all the projects also seem rather heavyweight and will be slightly harder to maintain in the long term. It feels like simply testing the "exclusion conditions" would suffice for testing the behavior with/without `*.ignore` files. 
4. We should double check again what the default behavior of without `.funcignore` should be -- I believe it should minimally ignore `local.settings.json` to be inline with function tools, see docs [here](https://learn.microsoft.com/en-us/azure/azure-functions/functions-run-local?tabs=macos%2Cisolated-process%2Cnode-v4%2Cpython-v2%2Chttp-trigger%2Ccontainer-apps&pivots=programming-language-java#upload-local-settings-to-azure).

Overall, I really do love the direction here. But I would love if we can have a simple clean landing of a minimal change -- there are currently 37 file changes in this PR alone. I'm more than happy to work with you to land a minimal change if we can document all the necessary requirements in a new issue referencing #1039 so we're all on the same page, and also to serve as official documentation for the feature.

Reply:
**On dotignore and recursive directories:**

Thanks for sharing your thoughts on dotignore. The current non-recursive behavior is consistent with how .funcignore works today, and I believe it's important to avoid introducing more differences. I’d prefer to maintain this alignment to keep things predictable and straightforward for developers.

**On changes to rzip.go:**

The changes in rzip.go are required to handle symbolic links in node_modules or virtual environments. Since these components could now be included, special code is necessary to ensure correct behavior. Keeping these updates within this PR ensures everything works seamlessly together.

**On e2e tests and test data files:**

Our current e2e tests are designed to cover any issues arising from the zip changes, particularly around symbolic links. The test data may be heavier, but it provides comprehensive coverage that ensures we don’t miss any potential edge cases.

**On default behavior for .funcignore:**

Regarding the behavior of .funcignore, I’m in touch with the functions team to verify if there are additional files that should be ignored, with or without the existence of a .funcignore file. It's also important to note that any issues where .funcignore doesn’t behave like .gitignore should be resolved within the func implementation. I don't think we should introduce behavior here that diverges from the expected .gitignore syntax.


Thank you for the offer, but I’m confident the current implementation is in line with our goals, and I don't feel any additional simplification is needed at this time. I appreciate your willingness to assist and your thoughtful feedback throughout this process

Reply:
I've left a comment on [rzip.go](https://github.com/Azure/azure-dev/pull/4258/files#r1775899423). It isn't clear what motivates all the changes here. I'd still push towards creating a separate issue for `rzip.go` change so that we can understand and track the changes separately.

> I’m in touch with the functions team to verify if there are additional files that should be ignored

Would we consider transferring all the existing documentation on the PR into an issue so that we can have the requirements discussion? I'm also more than happy to help here if needed.

Overall, I do really want this to see these changes landed -- I'm just trying to make sure we're covering enough our bases from enough angles to make this a good user experience for our users.

Reply:
Here are the list of default excludes for function service type:
https://github.com/Azure/azure-functions-core-tools/blob/37b826f143b99cd62308b88320e71aa806f6197f/src/Azure.Functions.Cli/Common/FileSystemHelpers.cs#L144

We may also want to confirm with app service team if there are default excludes like appsettings.development.json